superset/datasets/schemas.py

```diff
@@ -43,18 +43,12 @@
 
 
 def validate_python_date_format(value: str) -> None:
-    regex = re.compile(
-        r"""
-        ^(
-            epoch_s|epoch_ms|
-            (?P<date>%Y([-/]%m([-/]%d)?)?)([\sT](?P<time>%H(:%M(:%S(\.%f)?)?)?))?
-        )$
-        """,
-        re.VERBOSE,
-    )
-    match = regex.match(value or "")
-    if not match:
-        raise ValidationError([_("Invalid date/timestamp format")])
+    if value in ("epoch_s", "epoch_ms"):
+        return
+    try:
+        datetime.now().strftime(value or "")
```

Comment:
@mapledan would you mind using the [dateutil.parser.isoparse](https://dateutil.readthedocs.io/en/stable/parser.html#dateutil.parser.isoparse) method instead?

Reply:
Sure, I can use the `dateutil.parser.isoparse` method to make it support a more flexiable ISO 8601 format than the original regular expression validation.
However, let me confirm again, does python_date_format only support the ISO 8601 format?
Because based on the description I read, it is written as follows: "If the timestamp format does not adhere to the ISO 8601 standard, you will need to define an expression and type for transforming the string into a date or timestamp."
That's why I chose to use `strftime` for validation to handle the user's defined expression.

Reply:
@mapledan it should _only_ support ISO 8601—in addition to `epoch_s` and `epoch_ms`. The reason being is ISO 8601 guarantees lexigraphical ordering when comparing strings.

Reply:
Thanks for the explanation, the lexicographical ordering is crucial in data analysis.
In my test case, the `dateutil.parser.isoparse` strictly follows the ISO 8601 standard, so it does not allow YYYY/MM/DD format.
This will impact PR (#24113) as it allows the slash format.
And I traced the python_date_format.
It is using by [pandas.to_datetime](https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html) in `superset.utils.core.normalize_dttm_col`.

Which should we use as the validation, the`pandas.to_datetime` method, regex like `(?P<date>%Y([-/]?%m([-/]?%d)?)?)([\sT](?P<time>%H(:%M(:%S(\.%f)?)?)?))?`, or the `dateutil.parser.isoparse` method?

Reply:
Thanks @mapledan for flagging the other PR. Per the tooltip (see attached) it clearly states that the format should be ISO 8601, though the placeholder text uses `/`. Granted the `/` (in addition to `-`) does adhere to lexicographical ordering, but we should strictly adhere to the ISO 8601 standard for historical reasons.

<img width="434" alt="Screenshot 2023-11-17 at 8 48 07 AM" src="https://github.com/apache/superset/assets/4567245/877b3ced-93fe-46e4-a52b-a4d682c0f094">

@jfrag1 this PR reverts the logic you defined in https://github.com/apache/superset/pull/24113. I've also authored https://github.com/apache/superset/pull/26017 which updates the placeholder text which likely lead people astray.

Reply:
I think if we want to move towards strict ISO, we definitely need a migration to update existing `python_date_format`'s that currently use `/`.  Without a migration, any existing datasets with columns that use `/` will not be able to be updated without changing the `python_date_format` since they'll fail the validation.

Reply:
@jfrag1 a migration isn't possible as the string format is specified by the underly column in the dataset. Granted this change has the potential to churn users who were exposed to the relaxation in https://github.com/apache/superset/pull/24113, i.e., since 3.0 and thus they would need to update there dataset definition to use a SQL expression instead.

@michael-s-molina this is likely another great example of a breaking change and thus we likely need to wait until 4.0 to re-restrict the eligible formats.

Reply:
@john-bodley #24113 didn't expose a new relaxation to users; that PR was a follow-up to https://github.com/apache/superset/pull/23678, which changed the endpoint used to update dataset definitions.  The old endpoint did no validation at all for `python_date_format`, so there were people using slashes before that too (since the placeholder indicated to do so).

Reply:
Also I'm not sure what you mean by "a migration isn't possible", can't we have a migration that updates the table that stores column definitions?

Reply:
@jfrag1 the column in question isn't in the metadata database it's in the underlying analytical database(s), i.e., Trino, Hive, etc. where the temporal format is defined. It's difficult to estimate the impact of non-strict ISO 8601 validation, though since https://github.com/apache/superset/pull/24113 it only impacts newly registered datasets where the format is `%Y/%m/%d[...]`. Organizations likely have many registered temporal columns where the format is complete nonsense pre-https://github.com/apache/superset/pull/23678.

I think there's two option:

1. Relax requiring the ISO 8601 standard to include `/`, i.e., what is currently implemented (for right or wrong)*. This is potentially risky as the ISO 8601 standard _guarantees_ lexicographical ordering** whereas custom formatters don't.
2. Wait until 4.0 and mention in `UPDATING.md` the (re)restriction that only ISO 8601 formats are acceptable (enforced by way of validation at the API and database level) and that dataset owners will need to use a SQL expression instead to convert their string columns of the form `%Y/%m/%d` etc. to a `DATE`, `DATETIME`, etc. type.

\* Note due to lack of validation prior to https://github.com/apache/superset/pull/23678 (when we switched to the new RESTful API) there are likely organizations which have datasets which have a slew of formats in  `table_columns.python_datetime_format` column which don't adhere to the current format and thus wont work as expected. I was able to confirm this in Airbnb's Superset metadata database. As part of 4.0 we should add a database level CHECK constraint for the `table_columns.python_datetime_format` column which will force admins and dataset owners to correct any violating temporal columns.

\** Lexicographical ordering is essential when we're filtering temporal string columns via the `>`, `>=`, `<`, and `<=` operators.

This is a classic "shift left" problem where the underlying validation should also reside in the database layer. This is mentioned in [[SIP-99C] Proposal for model and business validation](https://github.com/apache/superset/issues/25828) and something that has been somewhat of an achilles heal for community over the years.